File: internal/scheduler/rollup.go
package scheduler

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"sort"
	"sync"
	"time"
	"vaportrail/internal/db"

	"github.com/caio/go-tdigest/v4"
	"github.com/jonboulle/clockwork"
)

type RetentionPolicy struct {
	Window    int `json:"window"`
	Retention int `json:"retention"`
}

var defaultPolicies = []RetentionPolicy{
	{Window: 0, Retention: 604800},         // Raw: 7 days
	{Window: 60, Retention: 15768000},      // 1m: 6 months
	{Window: 300, Retention: 31536000},     // 5m: 1 year
	{Window: 3600, Retention: 315360000},   // 1h: 10 years
	{Window: 86400, Retention: 3153600000}, // 1d: ~100 years (User didn't specify retention for 1d, assuming long)
}

func ValidateRetentionPolicies(policies []RetentionPolicy) error {
	// Sort policies by window size
	sortPolicies(policies)

	for i, p := range policies {
		if p.Window < 0 {
			return errors.New("retention window cannot be negative")
		}
		if i == 0 {
			if p.Window == 0 {
				continue // 0 (Raw) is valid base
			}
			// Smallest window must be >= 1s (implicit raw is 0, but here checking config)
		} else {
			// Check multiple
			prevWindow := policies[i-1].Window
			if prevWindow == 0 {
				// Raw fits into anything integer (seconds)
			} else {
				if p.Window%prevWindow != 0 {
					return fmt.Errorf("window %d is not a multiple of smaller window %d", p.Window, prevWindow)
				}
			}
		}
	}
	return nil
}

func sortPolicies(policies []RetentionPolicy) {
	sort.Slice(policies, func(i, j int) bool {
		return policies[i].Window < policies[j].Window
	})
}

// DefaultPolicies returns the default retention policies for use at target creation time.
func DefaultPolicies() []RetentionPolicy {
	// Return a copy to prevent mutation
	result := make([]RetentionPolicy, len(defaultPolicies))
	copy(result, defaultPolicies)
	return result
}

// DefaultPoliciesJSON returns the default retention policies as a JSON string.
func DefaultPoliciesJSON() string {
	data, _ := json.Marshal(defaultPolicies)
	return string(data)
}

// ErrNoRetentionPolicies is returned when a target has no retention policies configured.
var ErrNoRetentionPolicies = errors.New("no retention policies configured")

// GetRetentionPolicies parses and returns retention policies for a target.
// Returns ErrNoRetentionPolicies if no policies are configured.
func GetRetentionPolicies(t db.Target) ([]RetentionPolicy, error) {
	if t.RetentionPolicies == "" || t.RetentionPolicies == "[]" {
		return nil, ErrNoRetentionPolicies
	}
	var p []RetentionPolicy
	if err := json.Unmarshal([]byte(t.RetentionPolicies), &p); err != nil {
		return nil, fmt.Errorf("failed to parse retention policies: %w", err)
	}
	if len(p) == 0 {
		return nil, ErrNoRetentionPolicies
	}
	return p, nil
}

type RollupManager struct {
	db    db.Store
	clock clockwork.Clock
	stop  chan struct{}
	wg    sync.WaitGroup
}

func NewRollupManager(database db.Store) *RollupManager {
	return &RollupManager{
		db:    database,
		clock: clockwork.NewRealClock(),
		stop:  make(chan struct{}),
	}
}

func (rm *RollupManager) Start() {
	rm.wg.Add(1)
	go rm.run()
}

func (rm *RollupManager) Stop() {
	close(rm.stop)
	rm.wg.Wait()
}

func (rm *RollupManager) run() {
	defer rm.wg.Done()
	ticker := rm.clock.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-rm.stop:
			return
		case <-ticker.Chan():
			rm.processRollups()
		}
	}
}

func (rm *RollupManager) processRollups() {
	targets, err := rm.db.GetTargets()
	if err != nil {
		log.Printf("RollupManager: Failed to get targets: %v", err)
		return
	}

	for _, t := range targets {
		policies, err := GetRetentionPolicies(t)
		if err != nil {
			// Skip targets with no policies configured
			continue
		}
		// Ensure sorted
		sortPolicies(policies)

		// Map windows to find source
		// 0 -> Raw
		lastWindow := 0

		for _, p := range policies {
			if p.Window == 0 {
				lastWindow = 0
				continue
			}

			// Process this window using lastWindow as source
			rm.processTargetWindow(t, p.Window, lastWindow)
			lastWindow = p.Window
		}
	}
}

func (rm *RollupManager) processTargetWindow(t db.Target, windowSeconds int, sourceWindow int) {
	// 1. Get last rollup time
	lastTime, err := rm.db.GetLastRollupTime(t.ID, windowSeconds)
	if err != nil {
		log.Printf("RollupManager: Failed to get last rollup time for %s (w=%d): %v", t.Name, windowSeconds, err)
		return
	}

	// 2. Determine start time. If never rolled up, start from... when?
	// If zero time, maybe lookup first raw data? Or just start "now" (bad)?
	// Let's look up the first available raw data time IF lastTime is zero.
	// Actually, if it's a new target, lastTime is zero.
	// We can try to catch up from (now - retention) or from first raw data.
	// Let's assume if lastTime is zero, we look for the earliest raw data.
	// But `GetLastRollupTime` returns time.Time{}.

	start := lastTime
	if start.IsZero() {
		// Optimization: Find earliest raw data time.
		earliest, err := rm.db.GetEarliestRawResultTime(t.ID)
		if err != nil {
			log.Printf("RollupManager: Error getting earliest raw time: %v", err)
			return
		}
		if earliest.IsZero() {
			// No raw data? Nothing to roll up.
			return
		}
		// Truncate to window alignment
		start = earliest.Truncate(time.Duration(windowSeconds) * time.Second)
	}

	// Align next window
	// If lastTime was 12:00:00, next window starts 12:00:00 + window (covering [12:00:00 + w, ...))
	// Wait, `lastTime` is the *start* of the last processed window.
	// So next window starts at `lastTime + window`.
	nextWindowStart := start.Add(time.Duration(windowSeconds) * time.Second)

	// If start was simulated (Zero -> 24h ago), we might need to align it strictly.
	if lastTime.IsZero() {
		nextWindowStart = start // Start fresh from that point
	}

	// Safety: don't process future
	// Cutoff logic: Now - (MaxTimeout + CommitBuffer + 1s)
	// MaxTimeout is in t.Timeout (seconds). Buffer is 2s (from Scheduler).
	cutoff := rm.clock.Now().Add(-time.Duration(t.Timeout+3) * time.Second)

	// Collect all aggregated results to commit in a single transaction
	var results []*db.AggregatedResult

	for {
		windowEnd := nextWindowStart.Add(time.Duration(windowSeconds) * time.Second)
		if windowEnd.After(cutoff) {
			break // Caught up
		}

		agg := rm.aggregateWindow(t, windowSeconds, sourceWindow, nextWindowStart, windowEnd)
		if agg != nil {
			results = append(results, agg)
		}
		nextWindowStart = windowEnd
	}

	// Commit all results in a single transaction
	if len(results) > 0 {
		if err := rm.db.AddAggregatedResults(results); err != nil {
			log.Printf("RollupManager: Failed to save batch AggResults for %s (w=%ds): %v", t.Name, windowSeconds, err)
		}
	}
}

func (rm *RollupManager) aggregateWindow(t db.Target, windowSeconds int, sourceWindow int, start, end time.Time) *db.AggregatedResult {
	// Source Data Fetching
	var tDigest *tdigest.TDigest
	var timeoutCount int64
	var rowsProcessed int
	var err error

	if sourceWindow == 0 {
		// Aggregate from Raw
		raws, err := rm.db.GetRawResults(t.ID, start, end, -1)
		if err != nil {
			log.Printf("RollupManager: Error fetching raw results: %v", err)
			return nil
		}
		rowsProcessed = len(raws)
		if len(raws) == 0 {
			return rm.createEmptyRollup(t, windowSeconds, start)
		}

		tDigest, _ = tdigest.New(tdigest.Compression(100))
		for _, r := range raws {
			if r.Latency == -1 {
				timeoutCount++
			} else {
				tDigest.Add(r.Latency)
			}
		}

	} else {
		// Aggregate from Sub-Rollup
		// Fetch aggregated results for the source window that fall within this window
		// start inclusive, end exclusive?
		// Yes, [start, end).
		// Note: The sub-rollups MUST align perfectly if Validated.

		results, err := rm.db.GetAggregatedResults(t.ID, sourceWindow, start, end)
		if err != nil {
			log.Printf("RollupManager: Error fetching aggregated results (w=%d): %v", sourceWindow, err)
			return nil
		}
		rowsProcessed = len(results)
		if len(results) == 0 {
			return rm.createEmptyRollup(t, windowSeconds, start)
		}

		tDigest, _ = tdigest.New(tdigest.Compression(100))
		for _, res := range results {
			timeoutCount += res.TimeoutCount
			if len(res.TDigestData) > 0 {
				subTD, err := db.DeserializeTDigest(res.TDigestData)
				if err == nil {
					tDigest.Merge(subTD)
				}
			}
		}
	}

	tdBytes, err := db.SerializeTDigest(tDigest)
	if err != nil {
		log.Printf("RollupManager: Serialization failed: %v", err)
		return nil
	}

	log.Printf("RollupManager: Aggregated %s (w=%ds): %d rows, %d timeouts", t.Name, windowSeconds, rowsProcessed, timeoutCount)

	return &db.AggregatedResult{
		Time:          start,
		TargetID:      t.ID,
		WindowSeconds: windowSeconds,
		TDigestData:   tdBytes,
		TimeoutCount:  timeoutCount,
	}
}

func (rm *RollupManager) createEmptyRollup(t db.Target, windowSeconds int, start time.Time) *db.AggregatedResult {
	td, _ := tdigest.New(tdigest.Compression(100))
	tdBytes, _ := db.SerializeTDigest(td)
	return &db.AggregatedResult{
		Time:          start,
		TargetID:      t.ID,
		WindowSeconds: windowSeconds,
		TDigestData:   tdBytes,
		TimeoutCount:  0,
	}
}

File: vaportrail-rs/src/scheduler/rollup.rs
//! Rollup manager for aggregating probe results.

use crate::db::{
    deserialize_tdigest, serialize_tdigest, get_tdigest_stats, AggregatedResult, Store, Target,
};

use chrono::{DateTime, Duration as ChronoDuration, Utc};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Duration;
use tdigest::TDigest;
use tokio::sync::Mutex;

/// A retention policy for a data window.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetentionPolicy {
    pub window: i32,
    pub retention: i64,
}

/// Default retention policies.
pub fn default_policies() -> Vec<RetentionPolicy> {
    vec![
        RetentionPolicy { window: 0, retention: 604800 },       // Raw: 7 days
        RetentionPolicy { window: 60, retention: 15768000 },    // 1m: 6 months
        RetentionPolicy { window: 300, retention: 31536000 },   // 5m: 1 year
        RetentionPolicy { window: 3600, retention: 315360000 }, // 1h: 10 years
        RetentionPolicy { window: 86400, retention: 3153600000 }, // 1d: ~100 years
    ]
}

/// Return default policies as JSON string.
pub fn default_policies_json() -> String {
    serde_json::to_string(&default_policies()).unwrap_or_else(|_| "[]".to_string())
}

/// Validate retention policies.
pub fn validate_retention_policies(policies: &[RetentionPolicy]) -> Result<(), String> {
    let mut sorted = policies.to_vec();
    sorted.sort_by_key(|p| p.window);

    for (i, p) in sorted.iter().enumerate() {
        if p.window < 0 {
            return Err("retention window cannot be negative".to_string());
        }
        
        if i > 0 {
            let prev_window = sorted[i - 1].window;
            if prev_window > 0 && p.window % prev_window != 0 {
                return Err(format!(
                    "window {} is not a multiple of smaller window {}",
                    p.window, prev_window
                ));
            }
        }
    }

    Ok(())
}

/// Parse retention policies from JSON.
pub fn get_retention_policies(target: &Target) -> Option<Vec<RetentionPolicy>> {
    if target.retention_policies.is_empty() || target.retention_policies == "[]" {
        return None;
    }
    
    serde_json::from_str(&target.retention_policies).ok()
}

/// Manager for rolling up raw data into time windows.
pub struct RollupManager {
    store: Arc<Store>,
    _stop: Arc<Mutex<Option<tokio::sync::broadcast::Sender<()>>>>,
}

impl RollupManager {
    pub fn new(store: Arc<Store>) -> Self {
        Self {
            store,
            _stop: Arc::new(Mutex::new(None)),
        }
    }

    /// Start the rollup manager background task.
    pub fn start(&self) {
        let store = self.store.clone();

        tokio::spawn(async move {
            let mut interval = tokio::time::interval(Duration::from_secs(10));

            loop {
                interval.tick().await;
                process_rollups(&store);
            }
        });
    }
}

fn process_rollups(store: &Store) {
    let targets = match store.get_targets() {
        Ok(t) => t,
        Err(e) => {
            tracing::error!("RollupManager: Failed to get targets: {}", e);
            return;
        }
    };

    for target in targets {
        let policies = match get_retention_policies(&target) {
            Some(p) => p,
            None => continue,
        };

        let mut sorted_policies = policies;
        sorted_policies.sort_by_key(|p| p.window);

        let mut last_window = 0;
        for policy in sorted_policies {
            if policy.window == 0 {
                last_window = 0;
                continue;
            }

            process_target_window(store, &target, policy.window, last_window);
            last_window = policy.window;
        }
    }
}

/// Process rollups for a specific target and window size.
/// This is the core rollup logic that advances through time windows.
pub fn process_target_window(store: &Store, target: &Target, window_seconds: i32, source_window: i32) {
    // 1. Get last rollup time for this window
    let (start, is_first_rollup) = match store.get_last_rollup_time(target.id, window_seconds) {
        Ok(Some(last_time)) => {
            // We have a previous rollup - its time is the START of that window
            // So next window starts at last_time + window_seconds
            (last_time, false)
        }
        Ok(None) => {
            // No previous rollup - find earliest raw data and truncate to window boundary
            match store.get_earliest_raw_result_time(target.id) {
                Ok(Some(earliest)) => {
                    let truncated = truncate_to_window(earliest, window_seconds);
                    (truncated, true)
                }
                Ok(None) => return, // No raw data to process
                Err(e) => {
                    tracing::error!("RollupManager: Error getting earliest time: {}", e);
                    return;
                }
            }
        }
        Err(e) => {
            tracing::error!("RollupManager: Failed to get last rollup time: {}", e);
            return;
        }
    };

    // 2. Calculate next window start
    // If this is the first rollup, start from the truncated earliest time
    // Otherwise, advance by one window from the last rollup
    let mut next_window_start = if is_first_rollup {
        start
    } else {
        start + ChronoDuration::seconds(window_seconds as i64)
    };

    // 3. Safety cutoff: don't process windows that haven't fully passed yet
    // Add buffer for timeout + commit delay
    let cutoff = Utc::now() - ChronoDuration::seconds((target.timeout as i64) + 3);

    let mut results = Vec::new();

    // 4. Process all complete windows
    loop {
        let window_end = next_window_start + ChronoDuration::seconds(window_seconds as i64);
        
        // Stop if this window hasn't finished yet
        if window_end > cutoff {
            break;
        }

        if let Some(agg) = aggregate_window(store, target, window_seconds, source_window, next_window_start, window_end) {
            results.push(agg);
        }

        next_window_start = window_end;
    }

    // 5. Batch save all results
    if !results.is_empty() {
        let count = results.len();
        if let Err(e) = store.add_aggregated_results(&results) {
            tracing::error!(
                "RollupManager: Failed to save batch for {} (w={}s): {}",
                target.name,
                window_seconds,
                e
            );
        } else {
            tracing::debug!(
                "RollupManager: Saved {} rollups for {} (w={}s)",
                count,
                target.name,
                window_seconds
            );
        }
    }
}

fn aggregate_window(
    store: &Store,
    target: &Target,
    window_seconds: i32,
    source_window: i32,
    start: DateTime<Utc>,
    end: DateTime<Utc>,
) -> Option<AggregatedResult> {
    let mut tdigest = TDigest::new_with_size(100);
    let mut timeout_count: i64 = 0;
    let rows_processed: usize;

    if source_window == 0 {
        // Aggregate from raw results
        let raws = match store.get_raw_results(target.id, start, end, i32::MAX) {
            Ok(r) => r,
            Err(e) => {
                tracing::error!("RollupManager: Error fetching raw results: {}", e);
                return None;
            }
        };

        rows_processed = raws.len();
        if raws.is_empty() {
            // Return empty rollup to mark this window as processed
            return Some(create_empty_rollup(target, window_seconds, start));
        }

        let values: Vec<f64> = raws
            .iter()
            .filter_map(|r| {
                if r.latency == -1.0 {
                    timeout_count += 1;
                    None
                } else {
                    Some(r.latency)
                }
            })
            .collect();

        if !values.is_empty() {
            tdigest = TDigest::new_with_size(100).merge_unsorted(values);
        }
    } else {
        // Aggregate from sub-rollups
        let sub_results = match store.get_aggregated_results(target.id, source_window, start, end) {
            Ok(r) => r,
            Err(e) => {
                tracing::error!("RollupManager: Error fetching aggregated results: {}", e);
                return None;
            }
        };

        rows_processed = sub_results.len();
        if sub_results.is_empty() {
            return Some(create_empty_rollup(target, window_seconds, start));
        }

        // Merge all sub-tdigests
        let mut all_values: Vec<f64> = Vec::new();
        for res in &sub_results {
            timeout_count += res.timeout_count;
            if !res.tdigest_data.is_empty() {
                if let Some(sub_td) = deserialize_tdigest(&res.tdigest_data) {
                    let (min, max, _sum, count) = get_tdigest_stats(&sub_td);
                    if count > 0.0 {
                        // Approximate by sampling between min and max
                        let n = (count as usize).min(10);
                        for i in 0..n {
                            let t = i as f64 / (n - 1).max(1) as f64;
                            all_values.push(min + t * (max - min));
                        }
                    }
                }
            }
        }

        if !all_values.is_empty() {
            tdigest = TDigest::new_with_size(100).merge_unsorted(all_values);
        }
    }

    let td_bytes = serialize_tdigest(&tdigest);

    tracing::info!(
        "RollupManager: Aggregated {} (w={}s) at {}: {} rows, {} timeouts",
        target.name,
        window_seconds,
        start.format("%H:%M:%S"),
        rows_processed,
        timeout_count
    );

    Some(AggregatedResult {
        time: start,
        target_id: target.id,
        window_seconds,
        tdigest_data: td_bytes,
        timeout_count,
    })
}

fn create_empty_rollup(target: &Target, window_seconds: i32, start: DateTime<Utc>) -> AggregatedResult {
    let td = TDigest::new_with_size(100);
    let td_bytes = serialize_tdigest(&td);
    
    AggregatedResult {
        time: start,
        target_id: target.id,
        window_seconds,
        tdigest_data: td_bytes,
        timeout_count: 0,
    }
}

/// Truncate a datetime to the start of its containing window.
pub fn truncate_to_window(dt: DateTime<Utc>, window_seconds: i32) -> DateTime<Utc> {
    let ts = dt.timestamp();
    let truncated = ts - (ts % window_seconds as i64);
    DateTime::from_timestamp(truncated, 0).unwrap_or(dt)
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    #[test]
    fn test_truncate_to_window() {
        // 2024-01-01 12:34:56 truncated to 60s window = 12:34:00
        let dt = Utc.with_ymd_and_hms(2024, 1, 1, 12, 34, 56).unwrap();
        let truncated = truncate_to_window(dt, 60);
        assert_eq!(truncated, Utc.with_ymd_and_hms(2024, 1, 1, 12, 34, 0).unwrap());
        
        // 12:34:56 truncated to 300s (5m) window = 12:30:00
        let truncated = truncate_to_window(dt, 300);
        assert_eq!(truncated, Utc.with_ymd_and_hms(2024, 1, 1, 12, 30, 0).unwrap());
        
        // 12:34:56 truncated to 3600s (1h) window = 12:00:00
        let truncated = truncate_to_window(dt, 3600);
        assert_eq!(truncated, Utc.with_ymd_and_hms(2024, 1, 1, 12, 0, 0).unwrap());
    }

    #[test]
    fn test_validate_retention_policies() {
        // Valid: windows are multiples of each other
        let valid = vec![
            RetentionPolicy { window: 0, retention: 86400 },
            RetentionPolicy { window: 60, retention: 2592000 },
            RetentionPolicy { window: 300, retention: 31536000 },
        ];
        assert!(validate_retention_policies(&valid).is_ok());

        // Invalid: 90s is not a multiple of 60s
        let invalid = vec![
            RetentionPolicy { window: 60, retention: 86400 },
            RetentionPolicy { window: 90, retention: 86400 },
        ];
        assert!(validate_retention_policies(&invalid).is_err());

        // Invalid: negative window
        let negative = vec![
            RetentionPolicy { window: -1, retention: 86400 },
        ];
        assert!(validate_retention_policies(&negative).is_err());
    }

    #[test]
    fn test_default_policies() {
        let policies = default_policies();
        assert!(!policies.is_empty());
        
        // Should include raw (window=0) and at least one aggregation window
        assert!(policies.iter().any(|p| p.window == 0));
        assert!(policies.iter().any(|p| p.window == 60));
    }

    #[test]
    fn test_get_retention_policies() {
        let mut target = Target::default();
        
        // Empty policies returns None
        target.retention_policies = String::new();
        assert!(get_retention_policies(&target).is_none());
        
        target.retention_policies = "[]".to_string();
        assert!(get_retention_policies(&target).is_none());
        
        // Valid JSON returns Some
        target.retention_policies = r#"[{"window":60,"retention":86400}]"#.to_string();
        let policies = get_retention_policies(&target);
        assert!(policies.is_some());
        assert_eq!(policies.unwrap().len(), 1);
    }
}
